src/lib/Serialize.hs

```diff
@@ -104,10 +117,81 @@ prettyVal val = case val of
     return $ align $ group innerDoc
   atom -> return $ prettyPrec atom LowestPrec
 
+-- === taking memory snapshots for serializing heap-backed dex values ===
+
+data WithSnapshot a = WithSnapshot a [PtrSnapshot]  deriving Generic
+type RawPtr = Ptr ()
+-- TODO: we could consider using some mmap-able instead of ByteString
+data PtrSnapshot = ByteArray BS.ByteString
+                 | PtrArray [PtrSnapshot]
+                 | NullPtr  deriving Generic
+
+class HasPtrs a where
+  traversePtrs :: Applicative f => (PtrType -> RawPtr -> f RawPtr) -> a -> f a
+
+takeSnapshot :: HasPtrs a => a -> IO (WithSnapshot a)
+takeSnapshot x =
+  liftM (WithSnapshot x) $ execWriterT $ flip traversePtrs x \ptrTy ptrVal -> do
+    snapshot <- lift $ takePtrSnapshot ptrTy ptrVal
+    tell [snapshot]
+    return ptrVal
+
+restoreSnapshot :: HasPtrs a => WithSnapshot a -> IO a
+restoreSnapshot (WithSnapshot x snapshots) =
+  flip evalStateT snapshots $ flip traversePtrs x \_ _ -> do
+    (s:ss) <- get
+    put ss
+    lift $ restorePtrSnapshot s
+
+takePtrSnapshot :: PtrType -> RawPtr -> IO PtrSnapshot
+takePtrSnapshot _ ptrVal | ptrVal == nullPtr = return NullPtr
+takePtrSnapshot (_, ptrTy) ptrVal = case ptrTy of
+  PtrType eltTy -> do
+    childPtrs <- loadPtrPtrs ptrVal
+    PtrArray <$> mapM (takePtrSnapshot eltTy) childPtrs
+  _ -> ByteArray <$> loadPtrBytes ptrVal
+
+restorePtrSnapshot :: PtrSnapshot -> IO RawPtr
+restorePtrSnapshot snapshot = case snapshot of
+  PtrArray  children -> storePtrPtrs =<< mapM restorePtrSnapshot children
+  ByteArray bytes    -> storePtrBytes bytes
+  NullPtr            -> return nullPtr
+
+loadPtrBytes :: RawPtr -> IO BS.ByteString
```

Comment:
Honestly jumping up and down between all those functions has been really confusing for me. Pretty much all of them are used once, maybe let's just inline them?

Reply:
I actually prefer them as standalone functions (I think we differ on this style point). But I can at least reorder them so that they're closer to their use sites.